Add IHostBuilder support

Vandersteen opened the pull request:
As discussed here: https://github.com/getsentry/sentry-dotnet/issues/1001

First quick and dirty commit

My first thought after beginning to add an implementation: would it be bad if it acted 100% like the IWebHostBuilder extension.

The user would have the choice between:

```
Host.CreateDefaultBuilder(args)
    .ConfigureWebHostDefaults(webBuilder =>
    {
        webBuilder.UseSentry()
        webBuilder.UseStartup<Startup>();
    });
```

And

```
Host.CreateDefaultBuilder(args)
    .ConfigureWebHostDefaults(webBuilder =>
    {
        webBuilder.UseStartup<Startup>();
    })
    .UseSentry();
```

Only 'caveat' right now is that it would add a few classes that would never be used in a WorkerService environment:
* IStartupFilter, SentryStartupFilter
* IRequestPayloadExtractor, FormRequestPayloadExtractor
* IRequestPayloadExtractor, DefaultRequestPayloadExtractor

However i'm not sure this is such an issue, thoughts ?

Also, I haven't taken https://github.com/getsentry/sentry-dotnet/issues/1001#issuecomment-846674053 into consideration

Tyrrrz commented on src/Sentry.AspNetCore/SentryHostBuilderExtensions.cs:
```diff
@@ -0,0 +1,107 @@
+        /// <param name="builder">The builder.</param>
+        /// <returns></returns>
+        public static IHostBuilder UseSentry(this IHostBuilder builder)
+            => UseSentry(builder, (Action<SentryAspNetCoreOptions>?)null);
```
```suggestion
            => builder.UseSentry((Action<SentryAspNetCoreOptions>?)null);
```

Tyrrrz commented:
Can we converge on just one way of using `UseSentry()` somehow?

Vandersteen commented:
> Can we converge on just one way of using `UseSentry()` somehow?

What exactly do you mean ?

Tyrrrz commented:
> > Can we converge on just one way of using `UseSentry()` somehow?
> 
> What exactly do you mean ?

Sorry. I was referring to how having both the option to put `UseSentry()` inside `ConfigureWebHostDefaults(...)` and outside of it seems a bit confusing from user perspective. I wonder if we can do anything about that.

Vandersteen commented:
> > > Can we converge on just one way of using `UseSentry()` somehow?
> > 
> > 
> > What exactly do you mean ?
> 
> Sorry. I was referring to how having both the option to put `UseSentry()` inside `ConfigureWebHostDefaults(...)` and outside of it seems a bit confusing from user perspective. I wonder if we can do anything about that.

I'm not sure we can do anything about that.
IWebHostBuilder was the predecessor for IHostBuilder

I understand it could be confusing.

One way I suggested to 'tackle' that problem was explained in the original issue

* UseSentryWeb() on IWebHostBuilder
* UseSentryCore() on IHostBuilder

But that was also not desirable behaviour

Tyrrrz commented:
Let's see what @bruno-garcia thinks about this

bruno-garcia commented:
Not a fan of the two methods either.
Can we phase out the one on `IWebHostBuilder` then? Can we do everything through `IHostBuilder` instead?

Vandersteen commented:
My 2 cents, so you can make an informed decision

> Can we phase out the one on `IWebHostBuilder` then? 

That could indeed be an option, deprecate IWebHostBuilder extensions.
However, do note that `WebHost.CreateDefaultBuilder` will no longer be supported then, forcing users to use `Host.CreateDefaultBuilder`

While both still seem acceptable ways to create a Host from asp.net core's standpoint. There does not seem to be plans to deprecate/remove `WebHost`

```
// no longer supported
WebHost.CreateDefaultBuilder()
    .UseSentry()

// required
Host.CreateDefaultBuilder()
   .ConfigureWebHostDefaults(webBuilder =>
    {
        webBuilder.UseStartup<Startup>();
    })
    .UseSentry()
```

As far as I know, both of these are equivalent.
I don't see any signs that IWebHostBuilder will be deprecated as well (f.e. it is used in the callback of `ConfigureWebHostDefaults`)

>Can we do everything through `IHostBuilder` instead?

I can't see anything that you cannot achieve through the `IHostBuilder` that you can with the `IWebHostBuilder`

But don't take my word for it :)

bruno-garcia commented:
Sounds like we do need to keep both. So one way is: We add the infrastructure for `IHostBuilder` and update our docs to show only the use of this extension going forward.

We could add a note that `UseSentry` exists on `IWebHostBuilder` but is not needed if the one in `IHostBuilder` was already called.

All of that assumes that we can make them equivalent. If we can add "more" features through `IWebHostBuilder` this wouldn't really work.

Vandersteen commented:
Exactly, at the time of writing, the current `IHostBuilder` extensions in this PR are a 'literal' copy of the `IWebHostBuilder` with IWebHostBuilder replaced with IHostBuilder and some namespace / include changes.

bruno-garcia commented:
> Exactly, at the time of writing, the current `IHostBuilder` extensions in this PR are a 'literal' copy of the `IWebHostBuilder` with IWebHostBuilder replaced with IHostBuilder and some namespace / include changes.

Sounds like we could pull the code to a single place and forward the call from both places then?

Is the code reentrant already? We can expect that ppl will call it twice (from each interface).

Vandersteen commented:
> Sounds like we could pull the code to a single place and forward the call from both places then?

I looked into that, but IHostBuilder and IWebHostBuilder are not the same interface, so code sharing is not easily done

> Is the code reentrant already? We can expect that ppl will call it twice (from each interface).

No not yet, I will comment in code where I see issues

Vandersteen commented on src/Sentry.AspNetCore/SentryHostBuilderExtensions.cs:
```diff
@@ -0,0 +1,107 @@
+
+                _ = logging.Services
+                    .AddSingleton<IConfigureOptions<SentryAspNetCoreOptions>, SentryAspNetCoreOptionsSetup>();
+                _ = logging.Services.AddSingleton<ILoggerProvider, SentryAspNetCoreLoggerProvider>();
```
@bruno-garcia using TryAddSingleton here can cause issues, as it won't register it if let's say Serilog or other already exists, in this case you want the 'override' ability

bruno-garcia replied:
That's true. Is there a way to check if the specific specification has been added? I believe it was possible to check the items in the collection so we can verify if specifically `SentryAspNetCoreLoggerProvider` was already added and make sure we only add it once.

bruno-garcia commented on src/Sentry.AspNetCore/SentryHostBuilderExtensions.cs:
```diff
@@ -0,0 +1,107 @@
+            // The earliest we can hook the SDK initialization code with the framework
+            // Initialization happens at a later time depending if the default MEL backend is enabled or not.
+            // In case the logging backend was replaced, init happens later, at the StartupFilter
+            _ = builder.ConfigureLogging((context, logging) =>
```
The code inside thsi callback could be moved out and shared by `IHostBuilderExtensions` and `IWebHostBuilderExtensions`

Vandersteen replied:
Done

Vandersteen commented on src/Sentry.AspNetCore/SentryHostBuilderExtensions.cs:
```diff
@@ -0,0 +1,107 @@
+                    .AddSingleton<IConfigureOptions<SentryAspNetCoreOptions>, SentryAspNetCoreOptionsSetup>();
+                _ = logging.Services.AddSingleton<ILoggerProvider, SentryAspNetCoreLoggerProvider>();
+
+                _ = logging.AddFilter<SentryAspNetCoreLoggerProvider>(
```
There is no TryAddFilter, i'm not sure it is an issue if this is added twice thought

bruno-garcia replied:
Not sure either, I guess something we'd need to test?  There's no way to look into the filters already added?

Vandersteen replied:
Looks like it boils down to: https://github.com/aspnet/Logging/blob/2d2f31968229eddb57b6ba3d34696ef366a6c71b/src/Microsoft.Extensions.Logging/FilterLoggingBuilderExtensions.cs#L75

Vandersteen commented on src/Sentry.AspNetCore/SentryHostBuilderExtensions.cs:
```diff
@@ -0,0 +1,107 @@
+                configureSentry?.Invoke(context, sentryBuilder);
+            });
+
+            _ = builder.ConfigureServices((c, s) => _ = s.AddTransient<IStartupFilter, SentryStartupFilter>());
```
TryAddTransient is also an issue here, if any IStartupFilter was already added (which most likely it will) this will not be registered

bruno-garcia replied:
Something lke:

```
            if (!s.Any(d =>  d.ServiceType == descriptor.ServiceType &&    d.GetImplementationType() == implementationType))

```

Vandersteen replied:
Oh nice, I didn't think of that.
Just ran a quick test and yes that would work.

Vandersteen commented:
I dropped some comments here and there (see above)

In some code bases, I worked around this with a `private static bool` that keeps track if it is already called or not.
I could proceed and resolve those that way, so that calling it on both interfaces does not give strange issues

bruno-garcia commented:
> In some code bases, I worked around this with a private static bool that keeps track if it is already called or not.
> I could proceed and resolve those that way, so that calling it on both interfaces does not give strange issues

This strategy won't work in the ASP.NET Core case because even on a normal run of it, 2 containers are created and the first time we'd add (before the static fields are set) would end up on a dead host, and the new one wouldn't have it.
See: https://github.com/bruno-garcia/repro-logger-provider-double-instantiation https://github.com/getsentry/sentry-dotnet/issues/103#issuecomment-433882680 https://github.com/aspnet/Hosting/issues/1499

Vandersteen commented:
Took the comments into consideration, made  `AddSentry()` idempotent on

* IHostBuilder 
* IWebHostBuilder

bruno-garcia commented on src/Sentry.AspNetCore/Extensions/DependencyInjection/ServiceCollectionExtensions.cs:
```diff
@@ -20,17 +22,55 @@ internal static class ServiceCollectionExtensions
+        /// <returns></returns>
+        internal static IServiceCollection AddSentryStartupFilter(this IServiceCollection serviceCollection)
+        {
+            if(!serviceCollection.IsRegistered<IStartupFilter, SentryStartupFilter>())
```
nit: one day we'll have a linter that reformats and pushed to the branch :)

Could you please do this throughout the change?

```suggestion
            if (!serviceCollection.IsRegistered<IStartupFilter, SentryStartupFilter>())
```

bruno-garcia reviewed (commented):
I have just a nitpick but this already looks good to me!

Before we merge, It would be nice to have some coverage of this by some test, even an integration test of sorts.
I assume we have such test in the ASP.NET Core integration already then it could be copied over since it's the same thing.

It could also be used from one of the samples (Samples.ME.Logging or GenericHost actually)?
The samples make sure we have an easy way to run it and verify it works. And also to 'show ppl' some examples. And to have usage of the API to see what it looks like.

Vandersteen commented:
Fixed those `if (`'s I found 

I'll add some coverage and some samples

Vandersteen commented:
I had some trouble running it in .netcoreapp2.1 (IHostingEnvironment was unknown with HostBuilder)

I changed it to fix it, ran all (most) the aspnetcore samples, they still seem to work properly, however I can't check if the issues get to sentry

I changed the generic sample to use the new extension method.

Would you like me to make another sample for the Mvc's (using IHostBuilder instead), or shall I add it as a comment underneath ?

bruno-garcia commented:
Great stuff!

> Would you like me to make another sample for the Mvc's (using IHostBuilder instead), or shall I add it as a comment underneath ?

I wonder if we're just better off moving over to `IHostBuilder` on the `Sample.AspNetCore.Mvc` then. So we have one sample (`.Basic`) using `IWebHostBuilder` and the other using `IHostBuilder`. What do you think? 

Btw if you want to join us on Discord, we're online there often: http://discord.gg/sentry

Vandersteen commented:
hey @bruno-garcia i've tried to add some tests, but for some reason I can't seem to make them work.
I wasn't able to run these tests in debug mode either.

I keep getting: 

```
Diagnostic information:

Remaining (non-bound) argument specifications:
    any Action<IServiceCollection>

All argument specifications:
    any Action<IServiceCollection>

```

I don't understand how the `SentryWebHostBuilderTests` pass but not the `SentryHostBuilderTests`
I must be missing something silly.

Vandersteen commented:
Ok after a long debugging session, i finally found the issue.

IWebHostBuilder's interface contain
* ConfigureServices(Action<IServiceCollection> a);

While IHostBuilder does not, it has an extension method, that falls back to
* ConfigureServices(Action<HostBuilderContext, IServiceCollection> a)

Vandersteen commented:
@bruno-garcia @Tyrrrz Is there anything else I can do ?

Vandersteen commented:
Should've fixed failing tests

bruno-garcia commented on samples/Sentry.Samples.AspNetCore.Basic/Program.cs:
```diff
@@ -70,5 +73,21 @@ public static void Main(string[] args)
+                // .UseSentry("dsn") or .UseSentry(o => o.Dsn = ""; o.Release = "1.0"; ...)
+                .ConfigureWebHost(webBuilder =>
+                {
+                    webBuilder.UseStartup<Startup>();
```
What about this? Sometimes I find in samples it's good to be very explicit about things:

```suggestion
                    // Note that it's NOT needed to also call UseSentry here. That's because it's already hooked through IHostBuilder
                    webBuilder.UseStartup<Startup>();
```

bruno-garcia commented on samples/Sentry.Samples.AspNetCore.Basic/Startup.cs:
```diff
@@ -0,0 +1,70 @@
+using System;
```
I would prefer we remove this file. The idea of `Basic` is to have the simplest possible example.
I'd prefer we continued to use a single file for this example. We can even refactor this to use top level statements. 

We can remove the WebHost example and use only your new approach

Vandersteen replied:
What do you think of the current version of basic ?

bruno-garcia commented on samples/Sentry.Samples.GenericHost/Program.cs:
```diff
@@ -18,12 +18,8 @@ public static Task Main()
-                l.AddSentry();
-            })
+            .ConfigureLogging(b => b.AddConsole())
+            .UseSentry()
```
Very nice!

bruno-garcia commented on samples/Sentry.Samples.GenericHost/Sentry.Samples.GenericHost.csproj:
```diff
@@ -12,15 +12,15 @@
 
   <ItemGroup>
-    <ProjectReference Include="../../src/Sentry.Extensions.Logging/Sentry.Extensions.Logging.csproj" />
+    <ProjectReference Include="../../src/Sentry.AspNetCore/Sentry.AspNetCore.csproj" />
```
Do we need to change the dependency to ASP.NET Core integration? The new types are all on the S.E.L, right?
```suggestion
    <ProjectReference Include="../../src/Sentry.Extensions.Logging/Sentry.Extensions.Logging.csproj" />
```

Vandersteen replied:
New types are not on S.E.L. right now, see comment below

bruno-garcia commented on src/Sentry.AspNetCore/SentryHostBuilderExtensions.cs:
```diff
@@ -0,0 +1,93 @@
+        /// </summary>
+        /// <param name="builder">The builder.</param>
+        /// <returns></returns>
+        public static IHostBuilder UseSentry(this IHostBuilder builder)
```
`IHostBuilder` lives in `Microsoft.Extensions.Hosting` so this could be added to `Sentry.Extensions.Logging`.

Vandersteen replied:
I would've personally kept it in the asp.net core part, as it is the replacement of WebHostBuilder in the future.
Would you like me to move it ?

bruno-garcia reviewed (commented):
Thanks for this feature! There are a few minor things to address.
I'd even say it's likely good enough to merge and we can adjust things on our end. What would you say @Tyrrrz ?

Tyrrrz commented:
I would also prefer to avoid having a dependency on `Sentry.AspNetCore` when one needs a generic host integration. It seems counter-intuitive. Sentry.Extensions.Logging is in a weird place right now, in a sense that it should probably be called Sentry.Shared or something, but I think it's a better place for the generic `IHostBuilder` stuff.

Vandersteen commented:
I understand

How do we proceed with these dependencies:

* Adding the Startupfilter / middleware 
* Calling the AddSentry() on servicecollection extensions
* SentryAspNetCoreOptions
* Microsoft.AspNetCore.Hosting

Tyrrrz commented:
I'm also okay with merging as is since the PR has been up for a while, and then improving on it as we go along 😉

Vandersteen commented:
I'll leave that up to you guys

bruno-garcia commented:
> I understand
> 
> How do we proceed with these dependencies:
> 
> * Adding the Startupfilter / middleware
> * Calling the AddSentry() on servicecollection extensions (and it's depedencies on aspnetcore)
> * SentryAspNetCoreOptions
> * Microsoft.AspNetCore.Hosting

Could we move all of that to `Sentry.Extensions.Logging`?
Oh I see, we'd need to add `Microsoft.AspNetCore.Hosting` as a dependency or:

Instead of AspNetCore.Hosting, isn't this `Microsoft.Extensions.Hosting`?

```
  <ItemGroup Condition="'$(TargetFramework)' != 'netstandard2.0'">
    <FrameworkReference Include="Microsoft.AspNetCore.App" />
  </ItemGroup>
```

So it's either that or a new `Sentry.Extensions.Hosting` and have Sentry.AspNetCore depend on it?

Vandersteen commented:
I'll take a closer look tomorrow

Vandersteen commented:
Well I couldn't keep myself, took a quick gander.

Things like:
* IApplicationBuilder
* IStartupFilter
* HttpContext

All require AspNetCore ref

I believe you won't be able to substitute `IHostBuilder.UseSentry()` for `IWebHostBuilder.UseSentry()` completely if you go that path

`IHostBuilder.UseSentry()` could contain the minimum requirements in a `Sentry.Extensions.Hosting` but will not be a replacement for `IWebHostBuilder.UseSentry()`

I don't see anything wrong with that, if it's clear from the usage point of view.

Vandersteen commented:
@bruno-garcia Going forward, I think a new `Sentry.Extensions.Hosting` might be the best bet. Considering future .NET MAUI support

Do you wish for me to proceed on that path ?
Anything special I should know when setting up a new Project ?

bruno-garcia commented:
> @bruno-garcia Going forward, I believe a new `Sentry.Extensions.Hosting` might be the best bet. Considering future .NET MAUI support
> 
> Do you wish for me to proceed on that path ?
> Anything special I should know when setting up a new Project ?

Yeah sounds like we'll need it particularly because we're looking at adding support to MAUI as soon as it's out so we'd really appreciate your help with this!

Vandersteen commented:
Ok, thank you for the feedback.

I might have some time this week to continue working on this

mattjohnsonpint commented:
Thanks again for the time spent here, but we've decided not to do this.  See https://github.com/getsentry/sentry-dotnet/issues/190#issuecomment-1491065903

Thank you.
